modules/spacelift-automation/main.tf

```diff
@@ -0,0 +1,240 @@
+  #   ]
+  #   "spacelift-policies-notify-tf-completed" = [
+  #     "administrative",
+  #   ]
+  # }
+
+  _administrative_labels = {
+    for stack, configs in local.stack_configs : stack => ["administrative"] if tobool(try(configs.administrative, false)) == true
+  }
+
+  # Creates a map of `depends-on` labels for each stack based on the root module level dependency configuration.
+  # Example:
+  # {
+  #   "spacelift-spaces-clients" = [
+  #     "depends-on:spacelift-automation-mp-main",
+  #   ]
+  #   "spacelift-webhooks-notify-tf-completed" = [
+  #     "depends-on:spacelift-automation-mp-main",
+  #   ]
+  # }
+
+  _dependency_labels = {
+    for stack in local.stacks : stack => [
+      "depends-on:spacelift-automation-${terraform.workspace}"
+    ]
+  }
+
+  # Creates a map of folder labels for each stack based on Git structure for a proper grouping stacks in Spacelift UI.
+  # https://docs.spacelift.io/concepts/stack/organizing-stacks#label-based-folders
+  # Example:
+  # {
+  #   "spacelift-spaces-prod" = [
+  #     "folder:spacelift-spaces/prod",
+  #   ]
+  #   "spacelift-webhooks-notify-tf-completed" = [
+  #     "folder:spacelift-webhooks/notify-tf-completed",
+  #   ]
+  # }
+
+  _folder_labels = {
+    for stack in local.stacks : stack => [
+      "folder:${local.configs[stack].root_module}/${local.configs[stack].terraform_workspace}"
+    ]
+  }
+
+  # Merge all the labels into a single map for each stack.
+  # Example:
+  # {
+  #   "spacelift-spaces-clients" = [
+  #     "folder:spacelift-spaces/clients",
+  #     "administrative",
+  #   ]
+  # }
+
+  labels = {
+    for stack in local.stacks :
+    stack => compact(flatten([
+      lookup(local._administrative_labels, stack, []),
+      lookup(local._folder_labels, stack, []),
+      lookup(local._dependency_labels, stack, []),
+      try(local.stack_configs[stack].labels, []),
+    ]))
+  }
+
+  # Merge all before_init steps into a single map for each stack.
+  before_init = {
+    for stack in local.stacks : stack => compact(concat(
+      var.before_init,
+      try(local.stack_configs[stack].before_init, []),
+      # This command is required for each stack.
+      # It copies the tfvars file from the stack's workspace to the root module's directory
+      # and renames it to `spacelift.auto.tfvars` to automatically load variable definitions for each run/task.
+      ["cp tfvars/${local.configs[stack].terraform_workspace}.tfvars spacelift.auto.tfvars"],
+    )) if try(local.configs[stack].tfvars.enabled, false)
+  }
+}
+
+# Perform deep merge for common configurations and stack configurations
+module "deep" {
+  source   = "cloudposse/config/yaml//modules/deepmerge"
+  version  = "1.0.2"
+  for_each = local.root_module_stack_configs
+  # Stack configuration will take precedence and overwrite the conflicting value from the common configuration (if any)
+  maps = [local.common_configs[each.value.root_module], each.value]
+}
+
+resource "spacelift_stack" "this" {
```

Comment:
Mind expand why? I'm not opposed to `default`, just got used to `this` pattern.

Reply:
`this` says to me "one thing" and I think that is limiting. It's very JavaScript and is something that [I believe Anton B popularized in terraform-best-practices](https://www.terraform-best-practices.com/naming#resource-and-data-source-arguments).

I like `default` because it accomplishes the same thing that `this` is trying to do with "there is no descriptive name for this resource, so we need something", BUT it doesn't say to me that there is only one of those resources. You could have one `default` resource or many `default` resources, which is great because if we start off with a resource not using `for_each` or `count` then we can add it and avoid changing the resource name. 

Interestingly enough, I just went to see if Cloud Posse had a best practice around this [in their docs](https://docs.cloudposse.com/best-practices/terraform/), and it doesn't seem they do. 

Anyway, make sense? What do you think?

Reply:
Gotcha. Yeah, I think it's just so widely used (mainly because of Anton's modules) that it kinda became a pattern.
Your explanation makes sense to me 👍